reseq_ros2/reseq_ros2/communication.py

```diff
@@ -140,29 +163,34 @@ def ros_listener_callback(self, msg, module_num, topic_name):
             is_extended_id=True,
         )
         self.canbus.send(m)
-    
-    def topics_from_direction(self, d: rc.Direction): 
+
+    def topics_from_direction(self, d: rc.Direction):
         return list(filter(lambda x: x.direction == d, rc.topics))
-    
+
     def topic_from_id(self, id: int) -> rc.ReseQTopic:
         return next(filter(lambda x: x.can_id == id, rc.topics))
-    
+
     def topic_from_name(self, name: str) -> rc.ReseQTopic:
         return next(filter(lambda x: x.name == name, rc.topics))
 
 
 def main(args=None):
     rclpy.init(args=args)
+    communication = None
+    exit_code = 0  # Default to 0 for a clean exit
     try:
         communication = Communication()
-    except Exception as err:
-        print("Error while starting Communication node: " + str(err))
-        rclpy.shutdown()
-    else:
         rclpy.spin(communication)
-        communication.destroy_node()
-        rclpy.shutdown()
-
+    except Exception as err:
+        rclpy.logging.get_logger('communication').fatal(f"Error while starting Communication node: {str(err)}\n{traceback.format_exc()}")
+        exit_code = 1  # Set a non-zero exit code to indicate an error
+    finally:
+        if communication is not None:
+            exit_code = communication.exit_code if hasattr(communication, 'exit_code') else exit_code  # Use retcode if available
+            communication.destroy_node()
+        if rclpy.ok():
+            rclpy.shutdown()
+    sys.exit(exit_code)
```

Comment:
You approach works fine, however I believe that the one I linked you was easier to understand and more concise. 
```py
def main(args=None):
    rclpy.init(args=args)
    communication = None
    try:
        communication = Communication()
        rclpy.spin(communication)
    except Exception as err:
        rclpy.logging.get_logger('communication').fatal(f"Error while starting Communication node: {str(err)}\n{traceback.format_exc()}")
        raise err
    else:
        communication.destroy_node()
        rclpy.shutdown()
```

In my testings it worked, why did you chose to change the implementation from that one?

Reply:
Thank you for your feedback.

First and second comment respectively:
1. Do you mean comment changes by style changes? For example, `#Default config file path` to `# Default config file path` in the launch file?

2. I did this to fix the runtime error when the node crashes. For example, when I shut down the CAN bus, I also get `RuntimeError: Context must be initialized before it can be shutdown`. To solve this, I changed the implementation so that this misleading error doesn't occur anymore.

Reply:
Again, I will answer in order:
1. Yes, that is what I was talking about. 
2. The implementation suggested does not have that problem since it uses the `else` clause instead of the `finally` as in the previous implementation

Reply:
In the "near" future we will insert a formatter in CI/CD and this problem should be solved once for all

Reply:
I tried the implementation you linked. But it gave the runtime error that I mentioned when I tried to shut down the CAN bus. However, if you want me not to deal with it for now, I apply your implementation.

Reply:
Now I understand the problem (before I missed the monitoring function). I left some comments about that in the new review